README.md
# Questionmark::Barcodes

Questionmark::Barcodes (QM:B for short) was extracted from one of Questionmark's internal systems in order to share our custom barcode logic between multiple systems.

We've got some documentation on how we use [barcodes](./doc/barcodes.md) and [dummy EANs](./doc/dummy-eans.md), which led to the creation of this repository.

## Installation

Because this won't be published as a gem on Github, you can install it by adding this to the application's `Gemfile`:

    gem 'questionmark-barcodes', github: 'q-m/questionmark-barcodes', tag: 'v0.1.1'

This pins the gem to a specific release, allowing us to continue updating the repository and create new releases periodically that projects can adopt at their own pace.

This project intends to follow Semantic Versioning. See details below at "Releasing a new version".

## Usage

QM:B provides an extension to the GTIN classes implemented by [BarcodeValidation](https://github.com/marketplacer/barcodevalidation). Because of this, it works automatically:

```ruby
require 'questionmark-barcodes'
BarcodeValidation.scan("12345670") # => BarcodeValidation::GTIN::GTIN8
BarcodeValidation.scan("9669000000000") # => Questionmark::Barcodes::OldDummyGTIN
BarcodeValidation.scan("9669990132302229019") # => Questionmark::Barcodes::DummyGTIN
BarcodeValidation.scan("966999013200302000039007") # => Questionmark::Barcodes::FallbackDummyGTIN
```

As our custom GTINs start to offer more custom features, we'll add them here.

## Development

Useful commands after checking out the repo:

- `bin/setup` to install dependencies
- `rake spec` to run tests
- `rake rubocop` to run Rubocop, which checks the code style and may offer suggestions to improve the code
- `bin/console` for an interactive prompt that will allow you to experiment
- `rake ci` run all quality check tasks that CI will run, which is useful before you create a commit

We use GitHub for issues, pull requests and Continuous Integration. We don't publish our gem, but we do use version numbers for releases and tag them so our repositories can use code intended for release, rather than living on the bleeding edge.

## Releasing a new version

This project intends to follow [Semantic Versioning](https://semver.org/). This means the MAJOR version is incremented when we apply breaking changes. MINOR version increases are for new features that are backwards compatible. PATCH version increases are for backwards compatible bug fixes. While MAJOR version is ZERO, MINOR version signifies breaking/major changes and PATCH is incremented for all backwards compatible changes. Note that the stable branch on GitHub will be a work in progress, with unreleased changes.

- Check the [Changelog](./CHANGELOG.md) and ensure you've added a header for the new version + date that captures all unreleased changes.
- Update [version.rb](./lib/questionmark/barcodes/version.rb) to match this new version
- Create a local new branch "version-X-Y-Z" and commit the changes `git commit -am "Version X.Y.Z"`
- Run `bundle exec rake release` to create a tag based on this, and it and the commit them to GitHub. We automatically set the ENV variable `gem_push=no` in our [Rakefile](./Rakefile) to ensure we don't push the gem to rubygems.org.

## Contributing

Bug reports and pull requests are welcome on GitHub at:

- [Issues](https://github.com/q-m/questionmark-barcodes/issues)
- [Pull requests](https://github.com/q-m/questionmark-barcodes/pulls)

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## [Unreleased]

## [0.1.1] - 2022-11-11

The first _public_ release. Functionally unchanged from 0.1.0, but the documentation is friendlier.

- CI: trigger on pushes to `main` instead of `stable` ([#5](https://github.com/q-m/questionmark-barcodes/issues/5))
- CI: slightly changed name of task to indicate it's a CI task
- Doc: update descriptions to be more helpful for outside perspectives ([#7](https://github.com/q-m/questionmark-barcodes/issues/7))

## [0.1.0] - 2022-10-07

The initial release, offering feature parity with API's custom classes.

- Supply DummyGTIN, OldDummyGTIN and FallbackDummyGTIN as custom classes
- Added tests for all core BarcodeValidation GTIN classes and our own custom ones
- Added tests to prove GTINs with invalid check digit still parse as the correct class


spec/spec_helper.rb
# frozen_string_literal: true

require "questionmark/barcodes"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/questionmark/barcodes_spec.rb
# frozen_string_literal: true

RSpec.describe Questionmark::Barcodes do
  it "has a version number" do
    expect(Questionmark::Barcodes::VERSION).not_to be_nil
  end

  describe "BarcodeValidation extension" do
    # Valid GTINs and how we parse them
    {
      # Built-in GTINs from BarcodeValidation
      "12345670" => BarcodeValidation::GTIN::GTIN8,
      "123456789012" => BarcodeValidation::GTIN::GTIN12,
      "1234567890128" => BarcodeValidation::GTIN::GTIN13,
      "12345678901231" => BarcodeValidation::GTIN::GTIN14,
      # Our own custom GTINs
      # We started with length 13 custom GTINs
      "9669000000000" => Questionmark::Barcodes::OldDummyGTIN,
      # We have multiple types of length 19 dummy barcodes, based on their prefixes:
      # 96690* is a public dummy barcode
      "9669090132302229018" => Questionmark::Barcodes::DummyGTIN,
      # 96691* is a private dummy barcode
      "9669190132302229017" => Questionmark::Barcodes::DummyGTIN,
      # 966999* is a public dummy multi-product barcode
      "9669990132302229019" => Questionmark::Barcodes::DummyGTIN,
      # Any other 9669* GTIN of other lengths may have been unintentionally created, but is still supported this way
      "966999013200302000039007" => Questionmark::Barcodes::FallbackDummyGTIN
    }.each do |gtin, gtin_class|
      it "parses as #{gtin_class} when given valid GTIN #{gtin}" do
        parsed_gtin = BarcodeValidation.scan(gtin)
        expect(parsed_gtin.class).to eq(gtin_class)
        expect(parsed_gtin).to be_valid
      end

      # Transform our known valid GTINs to invalid ones in order to check they still parse to the correct class
      check_digit = gtin[-1, 1]
      wrong_check_digit = check_digit == "0" ? "9" : "0"
      invalid_gtin = gtin.dup
      invalid_gtin[-1, 1] = wrong_check_digit

      it "parses as #{gtin_class} when given invalid GTIN #{invalid_gtin}" do
        parsed_invalid_gtin = BarcodeValidation.scan(invalid_gtin)
        expect(parsed_invalid_gtin.class).to eq(gtin_class)
        expect(parsed_invalid_gtin).not_to be_valid
      end
    end
  end
end


lib/questionmark/barcodes/old_dummy_gtin.rb
# frozen_string_literal: true

require_relative "../barcodes"

module Questionmark
  module Barcodes
    # We use our own internal prefix for "dummy" barcodes.
    # This custom GTIN class supports our old length-13 dummy GTIN.
    #
    # @see lib/questionmark_dummy_gtin.rb for a broader overview of all our dummy GTINs.
    class OldDummyGTIN < BarcodeValidation::GTIN::Base
      # Any length 13 barcode will be evaluated by this before the default GTIN13, so prioritize it highly.
      prioritize_before BarcodeValidation::GTIN::GTIN13

      VALID_LENGTH = 13
      PREFIX = "9669"

      def self.handles?(input)
        input.start_with?(PREFIX) && input.length == VALID_LENGTH
      end

      def valid?
        self.class.handles?(input) && check_digit.valid?
      end
    end
  end
end


lib/questionmark/barcodes/fallback_dummy_gtin.rb
# frozen_string_literal: true

require_relative "../barcodes"

module Questionmark
  module Barcodes
    # We use our own internal prefix for "dummy" barcodes.
    # This custom GTIN class supports lengths that we don't explicitly know about,
    # such as the length 24 ones we've seen in specific cases.
    #
    # @see lib/questionmark_dummy_gtin.rb for a broader overview of all our dummy GTINs.
    class FallbackDummyGTIN < BarcodeValidation::GTIN::Base
      # This length is used for normalization but is _not_ the maximum we support.
      VALID_LENGTH = 19
      MIN_LENGTH = 6
      PREFIX = "9669"

      def self.handles?(input)
        input.start_with?(PREFIX) && input.length >= MIN_LENGTH
      end

      def valid?
        self.class.handles?(input) && check_digit.valid?
      end
    end
  end
end


lib/questionmark/barcodes/dummy_gtin.rb
# frozen_string_literal: true

require_relative "../barcodes"

module Questionmark
  module Barcodes
    # We use our own internal prefix for "dummy" barcodes.
    # These can be longer than regular EAN13s, but that's fine for our internal use.
    # This custom GTIN class supports the length 19 ones as a named first class citizen.
    #
    # Questionmark-specific dummy barcodes
    # For documentation about our dummy barcodes, see our documentation:
    # @see doc/dummy-eans.md
    class DummyGTIN < BarcodeValidation::GTIN::Base
      # The fallback supports length 6+ so ensure we are evaluated first
      prioritize_before FallbackDummyGTIN

      VALID_LENGTH = 19
      # We actually have a few sub-prefixes that we can extract separate classes for:
      # This is length 13, and handled by QuestionmarkOldDummyGTIN:
      # - 9669 - old dummy barcode
      #
      # These are length 19, and handled here:
      # - 96690 - public dummy barcode
      # - 96691 - private dummy barcode
      # - 966999 - public dummy multiproduct barcode
      PREFIX = "9669"

      def self.handles?(input)
        input.start_with?(PREFIX) && input.length == VALID_LENGTH
      end

      def valid?
        self.class.handles?(input) && check_digit.valid?
      end
    end
  end
end
